Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,738 @@
+            // (Section 5.4.1) of type PROTOCOL_ERROR.
+            throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+        }
+        
+        if flags.contains(.endHeaders) {
+            // no continuations following, so emit a full frame
+            frame.flags.formUnion(.endHeaders)
+            self.currentHeaderFrame = nil
+            return frame
+        } else {
+            self.currentHeaderFrame = frame
+            return nil
+        }
+    }
+}
+
+class HTTP2FrameEncoder {
+    private let allocator: ByteBufferAllocator
+    
+    init(allocator: ByteBufferAllocator) {
+        self.allocator = allocator
+    }
+    
+    /// Encodes the frame and optionally returns one or more blobs of data
+    /// ready for the system.
+    ///
+    /// Returned data blobs would include anything of potentially flexible
+    /// length, such as DATA payloads, header fragments in HEADERS or PUSH_PROMISE
+    /// frames, and so on. This is to avoid manually copying chunks of data which
+    /// we could just enqueue separately in sequence on the channel. Generally, if
+    /// we have a byte buffer somewhere, we will return that separately rather than
+    /// copy it into another buffer, with the corresponding allocation overhead.
+    ///
+    /// - Parameters:
+    ///   - frame: The frame to encode.
+    ///   - buf: Destination buffer for the encoded frame.
+    /// - Returns: An array containing zero or more additional buffers to send, in
+    ///            order. These may contain data frames' payload bytes, encoded
+    ///            header fragments, etc.
+    /// - Throws: Errors returned from HPACK encoder.
+    func encode(frame: HTTP2Frame, to buf: inout ByteBuffer) throws -> [IOData] {
+        // note our starting point
+        let start = buf.writerIndex
+        
+//      +-----------------------------------------------+
+//      |                 Length (24)                   |
+//      +---------------+---------------+---------------+
+//      |   Type (8)    |   Flags (8)   |
+//      +-+-------------+---------------+-------------------------------+
+//      |R|                 Stream Identifier (31)                      |
+//      +=+=============================================================+
+//      |                   Frame Payload (0...)                      ...
+//      +---------------------------------------------------------------+
+        
+        // skip 24-bit length for now, we'll fill that in later
+        buf.moveWriterIndex(forwardBy: 3)
+        
+        // 8-bit type
+        buf.write(integer: frame.payload.code)
+        // 8-bit flags -- we don't use padding when encoding in NIO, though
+        buf.write(integer: frame.flags.subtracting(.padded).rawValue)
+        // 32-bit stream identifier -- ensuring the top bit is empty
+        buf.write(integer: frame.streamID.safeNetworkStreamID ?? 0)
+        
+        func writePayloadSize(_ size: Int) {
```

Comment:
This probably belongs as a private extension on `ByteBuffer`.

Reply:
In my original variant I pulled out `_UInt24` and used that along with a `ByteBuffer` extension. Since I was only ever using it from one method in this implementation it seemed overkill to add it to `ByteBuffer`. Plus my OCD-ish tendencies wouldn't let me go without ensuring there's a parallel `write()` implementation, etc.

*twitches*

Reply:
Heh, I recommend suppressing the OCDish tendencies. This is a special case behaviour that we can just use Int with, rather than bring forward _UInt24.